### DISCUSSION TITLE
Toloka wants to be in providers

### DISCUSSION THREAD

**Discussion by Pocoder** [apache/airflow#21644] [Q&A]
Hello! I am from crowdsourcing tools group of [Toloka](https://toloka.ai/). 
We want to make helpful provider package. Now we have [this](https://github.com/Toloka/toloka-airflow). We are ready to actively enhance it. However, I have a couple of questions:
First, I see that there are no operators created with TaskFlowAPI in providers, why? 
Second, are our tasks enough generic or we should write something like [github](https://github.com/apache/airflow/blob/main/airflow/providers/github/operators/github.py) and [jira](https://github.com/apache/airflow/blob/main/airflow/providers/jira/operators/jira.py)?

**Comment by potiuk** [comment]
> First, I see that there are no operators created with TaskFlowAPI in providers, why?

Because TaskFlow API only works for DAGs. You Can't write Taskflow Operator. There is no such concept.

> Second, are our tasks enough generic or we should write something like 

Up to you. You are free to build and publish your own provider - even as a package. And it's totally up to you what functionality you expose there.

**Reply by Pocoder** [reply]
> Because TaskFlow API only works for DAGs. You Can't write Taskflow Operator. There is no such concept.

Airflow operators only works for DAGs too, aren't they? I can't see big difference between instantiating `PythonOperator` and calling function that decorated with `@task`. Am I right that the only problem when we create library of functions instead of classes is that clients of library are not able to pass specific parameters of operators such as `email_on_retry, email_on_failure, email_on_failure` etc. to our functions?

**Reply by potiuk** [reply]
No. Please read the taskflow docs. you can only use @task decorator in a DAG. But you can define Operatore elsewhere. You CAN add functions in your libraries that then you should be able to import in your own dag functions (and decorate the functions in dags). And you can publish your common functions in whatever way you want - this is python so you can do it. 

But you cannot make a library of @task decorated stuff because you can only add @task decorator in the context of DAG you write, not in the context of library you use.

In this context what you REALLY want - you want your "provider" to publish Hooks (which are already part of all providers). Then whenever somone wants to write taskflow DAG, they can use your Hooks  in it:

```
@task
def my_task():
    your_custom_hook = YourCustomHook()
    your_custom_hook.do_something().
```

But in this case you are not publishin "taskflow" operators. you are publishing Hooks thatn can be used if someone want to use it when they write a taskflow task. There is no "concept" of taskflow operator simply. It's always a "task" - decorated function that might reuse code from Hooks for example.

**Reply by Pocoder** [reply]
Thank you for your answers and sorry for my response delay. I found in [AIP 31](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=148638736) that @task decorator "Makes function an operator, but does not automatically assign it to a DAG (unless declared inside a  DAG context)". I created simple dag for experiment and I found out that function decorated with @task knows about dag even declared outside. It's cool because it makes @task decorated functions reusable. I understand that "taskflow operator" doesn't exists. However, it's possible to make library of @task decorated functions, isn't it?

```from airflow.decorators import dag, task
from airflow.operators.python import get_current_context
from airflow.utils.dates import days_ago


@task
def return_func_context():
    context = get_current_context()
    return context['dag'].dag_id


default_args = {
    'owner': 'airflow',
    'start_date': days_ago(5),
    'retries': 0,
}


@dag(default_args=default_args, schedule_interval=None, catchup=False, tags=['example_dags'])
def simple_dag():
    return_func_context()

dag = simple_dag()
```
It works fine and return 'simple_dag' into xcom

**Reply by potiuk** [reply]
Try it an let us know.

**Reply by Pocoder** [reply]
This approach works fine. However, we want to be in 'community maintained providers'. So, is it possible to get into there with package from our service (rewritten into operator classes if needed)? Or we should ask it at [Devlist](https://airflow.apache.org/community/)?

**Reply by potiuk** [reply]
I think we prefer that services that are not "vastly popular" are not maintained by the community. 

We are changing our approach and I think more and more new providers should be released and maintained by respective services  (and added to https://airflow.apache.org/ecosystem/#third-party-airflow-plugins-and-providers ) rather than contributed to Airflow code. I'd be rather against accepting new provider for Toloka to airflow community managed providers to be honest. but you can try devlist and raise the question.

**Reply by Pocoder** [reply]
Ok, thanks for your answers
